ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/InputsAndEffects.scala

```diff
@@ -85,25 +93,25 @@ private[kvutils] object InputsAndEffects {
           case fetch: Node.NodeFetch.WithTxValue[Value.ContractId] =>
             addContractInput(fetch.coid)
             fetch.key.foreach { keyWithMaintainers =>
-              inputs += globalKeyToStateKey(
-                GlobalKey(fetch.templateId, forceNoContractIds(keyWithMaintainers.key.value)))
+              addKeyInput(fetch.templateId, keyWithMaintainers)
             }
 
           case create: Node.NodeCreate.WithTxValue[Value.ContractId] =>
             create.key.foreach { keyWithMaintainers =>
-              inputs += globalKeyToStateKey(
-                GlobalKey(create.coinst.template, forceNoContractIds(keyWithMaintainers.key.value)))
+              addKeyInput(create.coinst.template, keyWithMaintainers)
             }
 
           case exe: Node.NodeExercises.WithTxValue[NodeId, Value.ContractId] =>
             addContractInput(exe.targetCoid)
+            exe.key.foreach { keyWithMaintainers =>
+              addKeyInput(exe.templateId, keyWithMaintainers)
+            }
```

Comment:
Does this work even if it's a regular consuming exercise on a contract that happens to have a key?

The tests imply as much, but I want to be sure.

Reply:
Yes, key is filled if there is a key, not if it's by key.